### PR TITLE
Windows Machine Config Operator design

### PR DESCRIPTION
**Pull request by aravindhp**
Enhancement proposal for the productization of Windows Containers that will enable running of Windows workloads on an OpenShift cluster.

### CODE REVIEW

File: enhancements/windows-containers/windows-machine-config-operator.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+The main motivation behind this enhancement is to satisfy customer
+requirement of being able to run Windows workloads on OpenShift clusters.
+
+### Goals
+
+As part of this enhancement we plan to do the following:
+* Provide workflows for adding Windows instances provisioned using the Machine
+  API to OpenShift clusters
+* Perform all the required steps within the VM for it to be added to the cluster
+  as an OpenShift worker node
```
Comment (elmiko) [inline enhancements/windows-containers/windows-machine-config-operator.md:50]: both these goals specifically mention "adding" to a cluster, but i am curious if this proposal is meant to cover the entire lifecycle of windows instances? (eg create and destroy)

i see the reference about not covering upgrading/downgrading, but i am more curious about the basic add/remove logic the Machine resources.

Comment (aravindhp) [reply enhancements/windows-containers/windows-machine-config-operator.md:50]: Removing a Windows worker node from the cluster will be achieved by scaling down the "Windows" MachineSet. LMK if you want me to call that out here.

Comment (elmiko) [reply enhancements/windows-containers/windows-machine-config-operator.md:50]: that sounds great. perhaps just change the wording a little in this text, eg:

"Provide workflows for adding and removing Windows instances ..."

"... VM for it to be added or removed from the cluster"

Comment (aravindhp) [reply enhancements/windows-containers/windows-machine-config-operator.md]: I have added your suggestion.

File: enhancements/windows-containers/windows-machine-config-operator.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+  as an OpenShift worker node and removed from it.
+
+### Non-Goals
+
+As part of this enhancement we do not plan to support:
+* Installing the container runtime in the Windows node
+* Upgrades (This will be addressed in a separate enhancement)
+  * Windows operating system upgrades
+  * Node management like k8s compute component upgrades, reboots and node
+    draining
```
Comment (enxebre) [inline enhancements/windows-containers/windows-machine-config-operator.md:59]: why is node draining not a goal?
Supporting draining should be just a consequence of the instance joining the cluster as nodes?
machine API blocks deletion on draining honouring PDBs by default.

File: enhancements/windows-containers/windows-machine-config-operator.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+will have all the pre-requisite steps needed for setting the VM up for key based
+SSH connections.
+
+When the cluster admin wants to bring up Windows worker nodes, they will create
+a [MachineSet](https://github.com/openshift/machine-api-operator/blob/master/config/machineset.yaml)
+where they reference a Windows OS image which has the Docker container runtime
+add-on enabled. They will also specify the user data secret created by WMCO as
+described above. WMCO will watch for Machine objects with a special label that
+indicate that they map to Windows VMs. Once the Machine object is in the
+"provisioned" state, WMCO will connect to the associated VM to transfer the
```
Comment (enxebre) [inline enhancements/windows-containers/windows-machine-config-operator.md:105]: s/state/phase. Can we include a link here https://github.com/openshift/enhancements/blob/master/enhancements/machine-api/machine-instance-lifecycle.md

File: enhancements/windows-containers/windows-machine-config-operator.md [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+* Provide workflows for adding and removing Windows instances provisioned using
+  the Machine API to OpenShift clusters
+* Perform all the required steps within the VM for it to be added to the cluster
+  as an OpenShift worker node and removed from it.
+
+### Non-Goals
+
+As part of this enhancement we do not plan to support:
+* Installing the container runtime in the Windows node
+* Upgrades (This will be addressed in a separate enhancement)
```
Comment (enxebre) [inline enhancements/windows-containers/windows-machine-config-operator.md:56]: what are the user story and expectations for a cluster that gets upgraded breaking the version compatibility policies between the API server and the kubelet running in a windows VM?

File: enhancements/windows-containers/windows-machine-config-operator.md (hunk 3 above)
Comment (aravindhp) [reply enhancements/windows-containers/windows-machine-config-operator.md:105]: Sure.

Comment (aravindhp) [reply enhancements/windows-containers/windows-machine-config-operator.md]: Addressed.

File: enhancements/windows-containers/windows-machine-config-operator.md (hunk 2 above)
Comment (aravindhp) [reply enhancements/windows-containers/windows-machine-config-operator.md]: Addressed

File: enhancements/windows-containers/windows-machine-config-operator.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+  * Windows operating system upgrades
+  * Node management like k8s compute component upgrades and reboots
+* OpenShift Builds
+* OpenShift DeploymentConfigs
```
Comment (sdodson) [inline enhancements/windows-containers/windows-machine-config-operator.md:60]: Just minor curiousity, what's unique about DeploymentConfigs that prevents their use on Windows?

File: enhancements/windows-containers/windows-machine-config-operator.md [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+When the cluster admin wants to bring up Windows worker nodes, they will create
+a [MachineSet](https://github.com/openshift/machine-api-operator/blob/master/config/machineset.yaml)
+where they reference a Windows OS image which has the Docker container runtime
+add-on enabled. They will also specify the user data secret created by WMCO as
+described above. WMCO will watch for Machine objects with a special label that
+indicate that they map to Windows VMs. Once the Machine object is in the
+["provisioned" phase](https://github.com/openshift/enhancements/blob/master/enhancements/machine-api/machine-instance-lifecycle.md),
+WMCO will connect to the associated VM to transfer the binaries and perform the
+following steps for it to join the cluster as a worker node:
+* Transfer the required binaries
```
Comment (sdodson) [inline enhancements/windows-containers/windows-machine-config-operator.md:107]: Drop this or how is this different from the mention of binaries being transferred two lines above?

File: enhancements/windows-containers/windows-machine-config-operator.md [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+* Configure the kubelet service using WMCB
+* Run the hybrid-overlay-node.exe binary
```
Comment (sdodson) [inline enhancements/windows-containers/windows-machine-config-operator.md:109]: Does this get installed as a persistent windows service or does WMCB have to start this each time the host boots?

File: enhancements/windows-containers/windows-machine-config-operator.md [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+This enhancement will start as GA.
+
+### Upgrade / Downgrade Strategy
+
+This is an involved topic and we will be opening a separate enhancement
+detailing it.
+
+### Version Skew Strategy
+
+We plan to maintain kubelet major version parity with the Linux counterpart.
```
Comment (sdodson) [inline enhancements/windows-containers/windows-machine-config-operator.md:155]: What about other binaries like hybrid-overlay-node.exe?

File: enhancements/windows-containers/windows-machine-config-operator.md [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+
+### Justification
+
+The reason we are using a WMCB binary rather than a container image is
+that all Windows container images are packaged with a Windows kernel and Red Hat
+has a policy to not ship 3rd party kernels for support reasons.
+
+### Design Details
+
+The WMCO image will be packaged with the following binaries:
```
Comment (sdodson) [inline enhancements/windows-containers/windows-machine-config-operator.md:80]: Everything below are windows binaries, right? Just want to make it clear these are artifacts that we copy to windows hosts.

**Comment by sdodson** [comment]
I'm fine with this as is to get us started. We have to ensure that we have an upgrades enhancement accepted before we consider windows containers to be GA so I'm looking forward to seeing the follow up enhancement for that.
/lgtm

File: enhancements/windows-containers/windows-machine-config-operator.md (hunk 9 above)
Comment (aravindhp) [reply enhancements/windows-containers/windows-machine-config-operator.md:80]: Correct, they are all Windows binaries that we copy to the Windows VM.

File: enhancements/windows-containers/windows-machine-config-operator.md (hunk 5 above)
Comment (aravindhp) [reply enhancements/windows-containers/windows-machine-config-operator.md:60]: From what I remember it is something about the intermediate containers that come up that launch the actual `Deployment`.

File: enhancements/windows-containers/windows-machine-config-operator.md (hunk 6 above)
Comment (aravindhp) [reply enhancements/windows-containers/windows-machine-config-operator.md:107]: You are correct. This is repeated.

File: enhancements/windows-containers/windows-machine-config-operator.md (hunk 7 above)
Comment (aravindhp) [reply enhancements/windows-containers/windows-machine-config-operator.md:109]: Hybrid-overlay at the moment cannot be configured as a Windows service and needs to be started each time the host boots. We have [WINC-295](https://issues.redhat.com/browse/WINC-295) and [WINC-296](https://issues.redhat.com/browse/WINC-296) that covers converting it into a Windows service.

File: enhancements/windows-containers/windows-machine-config-operator.md (hunk 8 above)
Comment (aravindhp) [reply enhancements/windows-containers/windows-machine-config-operator.md:155]: We will follow the same cadence with `hybrid-overlay-node.exe` also. I should have called that out here.

**Pull request merged**